BParkingNano/plugins/MuonTrackMerger.cc

```diff
@@ -0,0 +1,161 @@
+    tracks_ttracks_(consumes<TransientTrackCollection>(cfg.getParameter<edm::InputTag>("trackTransientTracks"))),
+    muonSelection_(cfg.getParameter<std::string>("muonSelection")),
+    trackSelection_(cfg.getParameter<std::string>("trackSelection")),
+    sortOutputCollections_(cfg.getParameter<bool>("sortOutputCollections"))
+{
+    produces<pat::CompositeCandidateCollection>("SelectedMuonsTracks");
+    produces<TransientTrackCollection>("SelectedTransientMuonsTracks");
+}
+
+  ~MuonTrackMerger() override {}
+
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions) {}
+
+
+private:
+  const edm::EDGetTokenT<pat::MuonCollection> muonsToken_;
+  const edm::EDGetTokenT<TransientTrackCollection> muons_ttracks_;
+  const edm::EDGetTokenT<pat::CompositeCandidateCollection> tracksToken_;
+  const edm::EDGetTokenT<TransientTrackCollection> tracks_ttracks_;
+
+  const StringCutObjectSelector<pat::CompositeCandidate> muonSelection_;
+  const StringCutObjectSelector<pat::CompositeCandidate> trackSelection_;
+
+  const bool sortOutputCollections_;
+};
+
+
+
+
+
+
+void MuonTrackMerger::produce(edm::StreamID, edm::Event &evt, edm::EventSetup const &stp) const {
+
+  //input
+  edm::Handle<pat::MuonCollection> muons;
+  evt.getByToken(muonsToken_, muons);
+
+  edm::Handle<TransientTrackCollection> muonsTT;
+  evt.getByToken(muons_ttracks_, muonsTT);
+
+  edm::Handle<pat::CompositeCandidateCollection> tracks;
+  evt.getByToken(tracksToken_, tracks);
+
+  edm::Handle<TransientTrackCollection> tracksTT;
+  evt.getByToken(tracks_ttracks_, tracksTT);
+
+
+
+  // output
+  std::unique_ptr<pat::CompositeCandidateCollection> muonTrack_out  (new pat::CompositeCandidateCollection);
+  std::unique_ptr<TransientTrackCollection> trans_muonTrack_out     (new TransientTrackCollection);
+  std::vector<float> pts;
+
+  auto muonSize = muons->size();
+  for(size_t mu_idx = 0; mu_idx < muonSize; ++mu_idx) {
+
+    const pat::Muon& mu = (*muons)[mu_idx];
+
+    pat::CompositeCandidate pcand;
+    pcand.setP4(mu.p4());
+    pcand.setCharge(mu.charge());
+    pcand.setVertex(mu.vertex());
+    pcand.setPdgId(mu.pdgId());
+    pcand.addUserInt("isPF", 1);
+    pcand.addUserInt("isTrack", 0);
+    pcand.addUserFloat("dxy", mu.dB(pat::Muon::PV2D));
+    pcand.addUserFloat("dxyS", mu.edB(pat::Muon::PV2D));
+    pcand.addUserFloat("dz", mu.dB(pat::Muon::PVDZ));
+    pcand.addUserFloat("dzS", mu.edB(pat::Muon::PVDZ));
+
+    if(!muonSelection_(pcand)) continue;
+
+    muonTrack_out->emplace_back(pcand);
+    trans_muonTrack_out->emplace_back(muonsTT->at(mu_idx));
+    pts.push_back(pcand.pt());
+  }
+
+  auto trkSize = tracks->size();
+  for(size_t trk_idx = 0; trk_idx < trkSize; ++trk_idx) {
+
+    const pat::CompositeCandidate& trk = (*tracks)[trk_idx];
+    if(!trackSelection_(trk)) continue;
+
+    pat::CompositeCandidate pcand;
```

Comment:
Tracks are already `pat::CompositeCandidate`, you can just do `pat::CompositeCandidate pcand(trk);`, it should copy the userFloats as well.